jellyfinclient/src/actions/ApiFunctions.js

```diff
@@ -0,0 +1,32 @@
+import ApiClient from 'jellyfin-apiclient/dist/apiclient';
+import jellyfinStore from '../utilities/storage/store';
+import * as types from './ActionTypes';
+
+export const connectToJellyfin = function (address) {
+    let apiClient = new ApiClient(null, address, "Jellyfin WebNG", '0.0.1', 'WebNG', 'WebNG', '');
+    jellyfinStore.store.dispatch({ type: types.UPDATE_APICLIENT, apiClient: apiClient });
+}
+
+export const loginToJellyfin = async function (username, password) {
+    let apiClient = copyClientFromStore();
+    let auth = await apiClient.authenticateUserByName(username, password);
+    jellyfinStore.store.dispatch(loginSuccessfully(auth.User.Name, auth.User.Id, auth.AccessToken));
+    apiClient.setAuthenticationInfo(auth.AccessToken, auth.User.Id);
+    jellyfinStore.store.dispatch({ type: types.UPDATE_APICLIENT, apiClient: apiClient });
+}
+
+export const copyClientFromStore = function () {
+    let apiClient = jellyfinStore.store.getState().jellyfinInterface.apiClient;
+    return Object.assign(new ApiClient(null, "-", "Jellyfin WebNG", "0.0.1", "WebNG", "WebNG", ""), apiClient);
```

Comment:
I'm not sure to understand why we need a clone here

Reply:
The Redux store is read only, therefore calling `setAuthenticationInfo` on the stored apiclient will fail. Cloning the Object, assigning the API-key and dispatching the new object is the "redux immutability" way to do it, I guess.

Reply:
Redux stores are not ment for storing objects. They are ment for state and serialization. At any time you should be able to export or import a state from a JSON object.

My goto approach to inject instances for object is usually with redux-thunk (which you already uses :))

`javascript
//INSIDE THE INDEX FILE OF THE PROJECT
import apiClient from './apiClient';

const services = {
    apiClient: new apiClient(); // This will create a single instance
};

let store = createStore(reducers, {},
    applyMiddleware(thunk.withExtraArgument(services)) //This will inject the services container in all redux actions
);

//MY REDUX ACTIONS LOOKS LIKE THIS
function myLoginAction() {
    return function (dispatch, getState, services) { //the services object is available as a 3rd param
        services.apiClient.loginAction((device) => {
            dispatch(addDevice(device));
        });
    }
}`

This would eliminate the clone

Reply:
I know many people are saying you should not store objects in redux storage, but I fail to understand redux thunk and it's benefits :laughing: . Feel free to change my code, or make an issue for that.

Reply:
No worries, once the PR is merged I might have a look at it. I don't want to drag the PR for nitti gritty things :)

Good job BTW.